content/blog/2019-10-23-divergence-optimizations/index.md

```diff
@@ -0,0 +1,231 @@
+
+While the coarse-grain specification of parallelization provides productivity, it also aggressively generates vector instructions. An obvious SPMD compilation procedure would generate a vector instruction for every instruction in the SPMD program. However, not every instruction needs to be parallelized. There are often values unknown at runtime, but constant across each parallel work-item. For example, memory indices and loop indices might be constant across lanes. In these cases, scalar instructions are optimal. A scalar instruction performs one operation for a group of parallel lanes and can later broadcast the single value to future vector instructions.
+
+## Target Architecture
+
+### Hardware
+
+We target a simplified version of the GPU architecture described in [IGC](https://dl.acm.org/citation.cfm?id=3314902). Each core contains a single ALU with a vector length of four as well as scalar, vector, and predicate register files. A SIMD instruction will use operands from the vector register and run on all lanes of the ALU. Predicate registers can be used to mask off certain lanes of the vector instructions when there is control flow. Scalar instructions will use a scalar register and only a single lane of the ALU. A scalar instruction and an equivalent vector instruction will complete in the same amount of time, but a scalar instruction will consume less energy than the vector instruction. Less dynamic energy is required: to (1) read a scalar register than a vector one and (2) use only a single lane of the ALU rather than every lane. We will save energy anytime we can replace a vector instruction with an equivalent scalar one.
+
+Although we target a specific architecture, every SIMD (Intel Integrated GPUs) and SIMT (nVidia and AMD Discrete GPUs) architecture can benefit from divergence optimizations.
+
+### ISA
+
+Predication was added to Bril to target the aforementioned hardware building upon the vector instructions added in our Project 1. A new instruction `vcmp` writes to a predicate register. A predicate register and its complement can be optionally specified before a vector instruction to mask off certain lanes of the ALU. Finally, two vector registers can be merged using the mask from a predicate register with a `vphi` instruction. An alternate implementation could remove the merge instruction and write to the same register directly at different indices. However, the former approach works better with the SSA format.
+
+```python
+# initialize vectors
+  va: vector = ...;
+  vb: vector = ...;
+
+# generate the predicate
+  p0: pred = vcmp va vb;
+
+# ignore lanes based on predicate mask
+  (p0)  vc0: vector = vadd va vb;
+# ignore lanes based on complement of predicate mask
+  (!p0) vc1: vector = vsub va vb;
+
+# merge lanes back together
+  vc: vector = vphi p0 vc0 vc1;
+
```

Comment:
Extra blank line. (This occurs a few more times below too, and it can make the rendered version look somewhat weird.)